src/client/http/transport.rs

```diff
@@ -1,237 +1,189 @@
-				.body(From::from(body))
-				.expect("Uri and request headers are valid; qed") // TODO: not necessarily true for URL here
-		});
-
-		Box::pin(async move {
-			let message = FrontToBack {
-				request: request.map_err(RequestError::Serialization)?,
-				send_back: self.responses_tx.clone(),
-			};
-
-			if requests_tx.send(message).await.is_err() {
-				log::error!("JSONRPC http client background thread has shut down");
-				return Err(RequestError::Http(Box::new(io::Error::new(
-					io::ErrorKind::Other,
-					"background thread is down".to_string(),
-				))));
-			}
+		let body = jsonrpc::to_vec(&request).map_err(|e| Error::Serialization(e))?;
 
-			Ok(())
-		})
+		if body.len() > self.max_request_body_size as usize {
+			return Err(Error::RequestTooLarge);
+		}
+
+		let req = hyper::Request::post(self.target.as_str())
+			.header(hyper::header::CONTENT_TYPE, hyper::header::HeaderValue::from_static(CONTENT_TYPE_JSON))
+			.header(hyper::header::ACCEPT, hyper::header::HeaderValue::from_static(CONTENT_TYPE_JSON))
+			.body(From::from(body))
+			.expect("URI and request headers are valid; qed");
+
+		let response = self.client.request(req).await.map_err(|e| Error::Http(Box::new(e)))?;
+
+		if response.status().is_success() {
+			Ok(response)
+		} else {
+			Err(Error::RequestFailure { status_code: response.status().into() })
+		}
 	}
 
-	/// Waits for the next response.
-	pub fn next_response<'s>(
-		&'s mut self,
-	) -> Pin<Box<dyn Future<Output = Result<jsonrpc::Response, RequestError>> + Send + 's>> {
-		Box::pin(async move {
-			let hyper_response = match self.responses_rx.next().await {
-				Some(Ok(r)) => r,
-				Some(Err(err)) => return Err(RequestError::Http(Box::new(err))),
-				None => {
-					log::error!("JSONRPC http client background thread has shut down");
-					return Err(RequestError::Http(Box::new(io::Error::new(
-						io::ErrorKind::Other,
-						"background thread is down".to_string(),
-					))));
-				}
-			};
-
-			if !hyper_response.status().is_success() {
-				return Err(RequestError::RequestFailure { status_code: hyper_response.status().into() });
-			}
+	/// Send notification.
+	pub async fn send_notification(&self, request: jsonrpc::Request) -> Result<(), Error> {
+		let _response = self.send_request(request).await?;
+		Ok(())
+	}
+
+	/// Send request and wait for response.
+	pub async fn send_request_and_wait_for_response(
+		&self,
+		request: jsonrpc::Request,
+	) -> Result<jsonrpc::Response, Error> {
+		let response = self.send_request(request).await?;
+		let body_size = read_content_length(response.headers()).unwrap_or(0);
+		let mut body_fut: hyper::Body = response.into_body();
+
+		if body_size > self.max_request_body_size {
+			return Err(Error::RequestTooLarge);
+		}
 
-			// Note that we don't check the Content-Type of the request. This is deemed
-			// unnecessary, as a parsing error while happen anyway.
+		let mut body = Vec::with_capacity(body_size as usize);
```

Comment:
The idea is that this will be faster for big payloads that needs be re-allocated but
I haven't benchmarked this change for bigger payloads because the [HttpServer](https://github.com/paritytech/jsonrpsee/blob/master/src/transport/http/server/background.rs#L243) hardcodes the max request size to 16384. 

Thus, it's just a guess.

Reply:
I like `with_capacity` even if it turns out it doesn't change performance at all (I suspect it doesn't and that the compiler does the right thing anyway) because it communicates to the reader that we intended to use this for the body before doing it.